File: Java/Copilot/NBodySimulationPar.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

public class NBodySimulationPar {
    private static final double G = 6.67430e-11; // Gravitational constant
    private static final double DT = 3600*24*365*1e-3; // Time step
    private static final int NUM_BODIES = 10000; // Number of bodies

    static class Body {
        double x, y, z;
        double vx, vy, vz;
        double mass;

        Body(double x, double y, double z, double vx, double vy, double vz, double mass) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx;
            this.vy = vy;
            this.vz = vz;
            this.mass = mass;
        }
    }

    private static void initializeBodies(List<Body> bodies) {
        bodies.add(new Body(0, 0, 0, 0, 0, 0, 1e30)); // Central body mass

        IntStream.range(1, NUM_BODIES).parallel().forEach(i -> {
            double angle = 2 * Math.PI * i / (NUM_BODIES - 1);
            bodies.add(new Body(
                Math.cos(angle) * 1e11,
                Math.sin(angle) * 1e11,
                0,
                -Math.sin(angle) * Math.sqrt(G * bodies.get(0).mass / 1e11),
                Math.cos(angle) * Math.sqrt(G * bodies.get(0).mass / 1e11),
                0,
                1e24 / NUM_BODIES // Small body mass
            ));
        });
    }

    // private static double calculateEnergy(List<Body> bodies) {
    //     return bodies.parallelStream().mapToDouble(body -> {
    //         double kinetic = 0.5 * body.mass * (body.vx * body.vx + body.vy * body.vy + body.vz * body.vz);
    //         double potential = bodies.stream().filter(other -> body != other).mapToDouble(other -> {
    //             double dx = body.x - other.x;
    //             double dy = body.y - other.y;
    //             double dz = body.z - other.z;
    //             double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
    //             return -G * body.mass * other.mass / distance;
    //         }).sum();
    //         return kinetic + 0.5 * potential;
    //     }).sum();
    // }

    private static void kickStep(List<Body> bodies) {
        bodies.parallelStream().forEach(body -> {
            double ax = 0.0, ay = 0.0, az = 0.0;
            for (Body other : bodies) {
                if (body != other) {
                    double dx = other.x - body.x;
                    double dy = other.y - body.y;
                    double dz = other.z - body.z;
                    double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
                    double force = G * other.mass / (distance * distance * distance);
                    ax += force * dx;
                    ay += force * dy;
                    az += force * dz;
                }
            }
            body.vx += ax * DT;
            body.vy += ay * DT;
            body.vz += az * DT;
        });

        bodies.parallelStream().forEach(body -> {
            body.x += body.vx * DT;
            body.y += body.vy * DT;
            body.z += body.vz * DT;
        });
    }

    public static void main(String[] args) {
        List<Body> bodies = new ArrayList<>(NUM_BODIES);
        initializeBodies(bodies);

        // double initialEnergy = calculateEnergy(bodies);
        // System.out.printf("Initial energy: %e%n", initialEnergy);

        for (int step = 0; step < 100; step++) {
            kickStep(bodies);
        }

        // double finalEnergy = calculateEnergy(bodies);
        // System.out.printf("Final energy: %e%n", finalEnergy);
    }
}

File: Julia/Copilot/NBodySimulationPar.jl
const G = 6.67430e-11 # Gravitational constant
const DT = 1e-3 # Time step
const NUM_BODIES = 1000000 # Number of bodies

struct Body
    x::Float64
    y::Float64
    z::Float64
    vx::Float64
    vy::Float64
    vz::Float64
    mass::Float64
end

function initialize_bodies()
    bodies = Vector{Body}(undef, NUM_BODIES)
    bodies[1] = Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1e30) # Central body mass

    for i in 2:NUM_BODIES
        angle = 2 * π * (i - 1) / (NUM_BODIES - 1)
        bodies[i] = Body(
            cos(angle) * 1e11,
            sin(angle) * 1e11,
            0.0,
            -sin(angle) * sqrt(G * bodies[1].mass / 1e11),
            cos(angle) * sqrt(G * bodies[1].mass / 1e11),
            0.0,
            1e24 # Small body mass
        )
    end
    return bodies
end

function calculate_energy(bodies)
    energy = 0.0
    Threads.@threads for i in 1:length(bodies)
        body = bodies[i]
        kinetic = 0.5 * body.mass * (body.vx^2 + body.vy^2 + body.vz^2)
        potential = 0.0
        for j in 1:length(bodies)
            other = bodies[j]
            if body != other
                dx = body.x - other.x
                dy = body.y - other.y
                dz = body.z - other.z
                distance = sqrt(dx^2 + dy^2 + dz^2)
                potential -= G * body.mass * other.mass / distance
            end
        end
        Threads.atomic_add!(energy, kinetic + 0.5 * potential)
    end
    return energy
end

function kick_step!(bodies)
    Threads.@threads for i in 1:length(bodies)
        body = bodies[i]
        ax, ay, az = 0.0, 0.0, 0.0
        for j in 1:length(bodies)
            other = bodies[j]
            if body != other
                dx = other.x - body.x
                dy = other.y - body.y
                dz = other.z - body.z
                distance = sqrt(dx^2 + dy^2 + dz^2)
                force = G * other.mass / (distance^3)
                ax += force * dx
                ay += force * dy
                az += force * dz
            end
        end
        body.vx += ax * DT
        body.vy += ay * DT
        body.vz += az * DT
    end

    Threads.@threads for i in 1:length(bodies)
        body = bodies[i]
        body.x += body.vx * DT
        body.y += body.vy * DT
        body.z += body.vz * DT
    end
end

function main()
    bodies = initialize_bodies()

    initial_energy = calculate_energy(bodies)
    println("Initial energy: $initial_energy")

    for step in 1:1000
        kick_step!(bodies)
    end

    final_energy = calculate_energy(bodies)
    println("Final energy: $final_energy")
end

main()
